Mixed Models!

FBartos opened the pull request:
Known bugs:
- changing UI scale destroys contrast widget

Unit tests are half-done and not updated since they are supposed to be changed in the future.

required files
https://github.com/FBartos/jasp-required-files/tree/Windows-MM

TimKDJ commented:
Great to see you finished it!

I will give this a more in-depth review soon (there are a couple of PR's I need to get through first), but in the mean time please reformat the code according to our [JASP style guide](https://github.com/jasp-stats/jasp-desktop/blob/stable/Docs/development/r-style-guide.md) (it should suffice to just do ctrl+a and then ctrl+shift+a in RStudio).

FBartos commented:
ok, done. I just wanna state that I don't like this formating and I find it gross.

TimKDJ commented:
Note that RStudio's beautifier is very stringent with respect to 80 (?) char lines. As you can read in the style guide, we are not. But I'm not sure if RStudio can be modified to behave differently (@vandenman?). You can also make the style guide changes manually.

FBartos commented:
well, let's hope that I won't have to look at the code ever again .)

vandenman commented:
You can adjust the margin to e.g., 120 characters? Tools -> Global Options -> Code -> Display -> Show margin -> margin. I'm not sure if that also affects the auto-indent though.

Also, when I briefly looked at the code I noticed that you use `&` rather than `&&` inside of if statements (same for `|` and `||`). Inside if statements, you should always use double ampersands: `&&`. See also ``?`&&` ``.

Finally,

> well, let's hope that I won't have to look at the code ever again.

By writing it you're now also the maintainer :-)

FBartos commented:
> Also, when I briefly looked at the code I noticed that you use & rather than && inside of if statements (same for | and ||). Inside if statements, you should always use double ampersands: &&. See also ?`&&` .
I can change that to ||/&& but I think that there is a point of using single ones since it throws a warning when the argument doesn't have the expected length.

> By writing it you're now also the maintainer :-)
I know, If the code were to be spotless, then there would be no need to edit it in the future. It's probably, p<.05, not gonna be the case though. :D

vankesteren commented:
Hey! I was asked to review, but I can't seem to get it running. What do I need here? Which branch of jasp-desktop? Do I just load it as a developer module or is it possible to get a .tar.gz file for this module?

vandenman commented:
Same for me. I'm not working on Windows, @FBartos can you provide a list of packages that I need to install?

FBartos commented:
Hi guys, you need a stable branch of jasp-desktop, and either [this ](https://github.com/jasp-stats/jasp-required-files/tree/windows-MixedModes-and-RoBMA)branch of required files (for Windows) or you can install these packages manually.

`
pckgs <- c("bayesplot","lme4", "rstanarm", "shinystan", "dygraphs", "threejs", "crosstalk", "crosstalk", "shinythemes", "extraDistr", "ggpol", "beeswarm", "ggbeeswarm",  "vipor", "ggridges", "DT", "rsconnect", "statmod")

devtools::install_github("https://github.com/bayesstuff/stanova", lib = "....")
devtools::install_github("https://github.com/singmann/afex",      lib = "....")
devtools::install_github("https://github.com/rvlenth/emmeans",    lib = "...")
`

You can install the package as a developer module, or there is a .tar.gz version.
[JASP-MixedModels.tar.gz](https://github.com/jasp-stats/MixedModels/files/4721588/JASP-MixedModels.tar.gz)
Thanks!

FBartos commented:
btw, the GLMMs will probably fail to converge in many cases if you use the debug dataset, since the variables are far from optimal for fitting there.

vandenman commented:
> I can change that to ||/&& ...

Please do. 

From the helpfile, 

> & and && indicate logical AND and | and || indicate logical OR. The shorter form performs elementwise comparisons in much the same way as arithmetic operators. The longer form evaluates left to right examining only the first element of each vector. Evaluation proceeds only until the result is determined. The longer form is appropriate for programming control-flow and typically preferred in if clauses.

The creators of R recommend `&&` in if statements and we do that everywhere in JASP. The has the benefit that whenever one expression evaluates to FALSE in `&&` (or TRUE in `||`) the other expressions are not evaluated.

vandenman commented:
>  pckgs <- c("bayesplot","lme4", "rstanarm", "shinystan", "dygraphs", "threejs", "crosstalk", "crosstalk", "shinythemes", "extraDistr", "ggpol", "beeswarm", "ggbeeswarm", "vipor", "ggridges", "DT", "rsconnect", "statmod")

These are a ton of dependencies, are these really all necessary? I notice that they're only mentioned under `Suggests` on CRAN.

vankesteren commented:
yeah I agree, that's a _lot_ of (unnecessary?) dependencies!

vankesteren commented:
also for some reason I can't checkout the `Windows-MM` branch...???
![afbeelding](https://user-images.githubusercontent.com/11596858/83609399-8a1f7080-a57e-11ea-9c90-b9fff82a016f.png)

vankesteren commented:
This will be the very first feature request by users: __support for wide data format__

FBartos commented:
> The creators of R recommend && in if statements and we do that everywhere in JASP. This has the benefit that whenever one expression evaluates to FALSE in && (or TRUE in ||) the other expressions are not evaluated.

OK, I changed that.

> These are a ton of dependencies, are these really all necessary? I notice that they're only mentioned under Suggests on CRAN.
> 
> In particular, shinythemes should not be mandatory right?

We have quite a lot different plotting options for ggplot and I wasn't able to run it without these packages.

> also for some reason I can't checkout the Windows-MM branch...???

I don't really understand git and this error message. Could it be due to the fact that the branch is on my fork? Otherwise, I linked a different branch that's part of jasp-required-files

> This will be the very first feature request by users: support for wide data format

Then there is the repeated measures ANOVA that can do that. I think that we would need standalone feature in JASP to do that conversion. It would complete mess the whole input menu otherwise IMO.

vankesteren commented:
Mixed models crashes on my machine with the following file (from `lme4` datasets). After adding "day" as a predictor, the program just quits. Also, mixed models is available even without data??

Installed on the stable branch using your tar.gz file, requires files is on windows mixed models branch

[sleepstudy.zip](https://github.com/jasp-stats/MixedModels/files/4721832/sleepstudy.zip)

![](https://im5.ezgif.com/tmp/ezgif-5-d5f836a4280b.gif)

vankesteren commented on R/MixedModelsCommon.R:
```diff
@@ -563,761 +794,1135 @@
+          Inf)
+    }
+    
+    # TODO: deal with the emtrends scoping problems
```
why do you need to put this stuff in the parent environment?

FBartos replied:
I'm not sure what do you mean by that. I found this way of passing this two arguments in emmeans documentation

vandenman commented:
Okay, so the four functions `MixedModelsLMM`, `MixedModelsBLMM`, `MixedModelsBGLMM`, and `MixedModelsGLMM` are remarkably similar. Did you consider making one function that takes a string for dispatching? I like how you dispatch based on the different methods inside each subfunction and I think you could take it even one step further (like in the Bayesian ANOVAs for example).

Now I wonder why the [`LMM`](https://github.com/FBartos/JASP-MixedModels/blob/f98a8af134e47cdeaabe10a1268a2b8caab00e99/R/MixedModelsLMM.R#L42) has
```r
  if (length(options$plotsX) > 0 &&
      is.null(jaspResults[["plots"]]))
        .mmPlot(jaspResults, dataset, options)

```
while the [`GLMM`](https://github.com/FBartos/JASP-MixedModels/blob/f98a8af134e47cdeaabe10a1268a2b8caab00e99/R/MixedModelsGLMM.R#L42) has
```r
  if (length(options$plotsX))
    .mmPlot(jaspResults, dataset, options, "GLMM")
```

vankesteren commented on R/MixedModelsCommon.R:
```diff
@@ -1820,142 +2640,250 @@
-.mmMessageResponse    <- "Results are on the response scale." 
-.mmMessageNotResponse <- "Results are not on the response scale and might be misleading." 
-.mmMessageANOVAtype     <- function(type){
+.mmMessageInterpretability <-
+  "The intercept corresponds to the (unweighted) grand mean and for each factor with k levels k - 1 parameters are added. Thus, estimates shown here cannot be mapped 1 to 1 to factors with more than two levels (as no such mapping exists). For factors with two levels, corresponding estimates represent half the difference between the factor levels, for factors with more than two levels estimates sh ...
+.mmMessageSingularFit <-
+  "Model fit is singular. Specified random effects parameters (random intercepts and random slopes) cannot be estimated from the available data. Carefully reduce the random effects structure, but be aware this may induce unknown risks of anti-conservative results (i.e., p-values might be lower than nominal)."
+.mmMessageVovkSellke <-
+  "Vovk-Sellke Maximum <em>p</em>-Ratio: Based on a two-sided <em>p</em>-value, the maximum possible odds in favor of H\u2081 over H\u2080 equals 1/(-e <em>p</em> log(<em>p</em>)) for <em>p</em> \u2264 .37 (Sellke, Bayarri, & Berger, 2001)."
+.mmMessageNumericalProblems <-
+  "Numerical problems with the maximum-likelihood estimate (e.g., gradients too large). This may indicate that the specified random effects parameters (random intercepts and random slopes) cannot be estimated from the available data. Consider carefully reducing the random effects structure, but be aware this may induce unknown risks of anti-conservative results (i.e., p-values might be lower than ...
+.mmMessageDFdisabled <-
+  "Estimation of degrees of freedom disabled (i.e., asymptotic results shown), because the number of observations is large. To force estimation, check corresponding option."
+.mmMessageResponse    <- "Results are on the response scale."
+.mmMessageNotResponse <-
+  "Results are not on the response scale and might be misleading."
```
So here again missing `gettext()`

FBartos replied:
yes, I will update that

FBartos commented:
> Mixed models crashes on my machine with the following file (from lme4 datasets). After adding "day" as a predictor, the program just quits.

Hmm, I was able to replicate that behavior using nightly (that has the proper set-up), but I was getting inconsistent results with my built JASP. It either crashed the JASP upon loading the csv file (with no output to the qml console) itself or proceeded without a problem.

> Also, mixed models is available even without data??

no, I forgot to remove that from testing it

vankesteren commented on inst/help/MixedModelsBGLMM.md:
```diff
@@ -0,0 +1,149 @@
+Bayesian Generalized Linear Mixed Models allow you to model a linear relationship between one or more explanatory variable(s) and a continuous dependent variable in cases where the observations are not independent, but clustered given one or several random effects grouping factors (e.g., repeated measures across participants or items, children within schools). They are generalization of Bayesian  ...
+
+### Assumptions
+- Linearity and additivity: The response variable is linearly related to all predictors and the effects of the predictors are additive on the linear scale.
```
minor comment: the response variable does not have to be linearly related, there is a link function involved too!

vankesteren commented on inst/help/MixedModelsBGLMM.md:
```diff
@@ -0,0 +1,149 @@
+- Linearity and additivity: The response variable is linearly related to all predictors and the effects of the predictors are additive on the linear scale.
+- Independence of errors: The errors are uncorrelated with each other after taking the model (i.e., fixed effects and random effects structure) into account.
+- Homoscedasticity: The error variance of each predictor is constant across all values of that predictor.
+- Normality of errors: The errors are normally distributed with mean zero.
```
??? this is _Generalized_ linear mixed models right?

vankesteren commented on inst/help/MixedModelsBGLMM.md:
```diff
@@ -0,0 +1,149 @@
+  - Distribution function which likelihood will be used for the dependent variable.
+
+#### Link
+  - Link function that will be used to the mean parameter of the selected distribution function.
```
used to ___ the mean

vankesteren commented on inst/help/MixedModelsBGLMM.md:
```diff
@@ -0,0 +1,149 @@
+- Estimated grand mean and estimated deviations from the grand mean for all levels of each fixed effects model term. It can be changed to estimated marginal means in the `Options` section.
+  - Level: Levels of the fixed effects model term.
+  - Estimate: This column contains the independent variables or their interaction.
+  - SE: Standart error of the estimate.
```
standard

FBartos commented:
> Okay, so the four functions `MixedModelsLMM`, `MixedModelsBLMM`, `MixedModelsBGLMM`, and `MixedModelsGLMM` are remarkably similar. Did you consider making one function that takes a string for dispatching? I like how you dispatch based on the different methods inside each subfunction and I think you could take it even one step further (like in the Bayesian ANOVAs for example).

I see I will change it appropriately.

> Now I wonder why the ...

Yeah, I missed that while updating the code, thanks. I will unify that as well.

vankesteren commented on inst/help/MixedModelsGLMM.md:
```diff
@@ -0,0 +1,152 @@
+Generalized Linear Mixed Models allow you to model a linear relationship between one or more explanatory variable(s) and a continuous dependent variable in cases where the observations are not independent, but clustered given one or several random effects grouping factors (e.g., repeated measures across participants or items, children within schools). They are generalization of Linear Mixed Model ...
+
+### Assumptions
+- Linearity and additivity: The response variable is linearly related to all predictors and the effects of the predictors are additive on the linear scale.
+- Independence of errors: The errors are uncorrelated with each other after taking the model (i.e., fixed effects and random effects structure) into account.
+- Homoscedasticity: The error variance of each predictor is constant across all values of that predictor.
+- Normality of errors: The errors are normally distributed with mean zero.
```
same thing here as with bayesian: linearity only with link function, and errors are distributed according to distributional family (link-inverse linear combination of the predictors is the expectation of this distribution)
y ~ DistributionalFamily(linkinv(x*b), dispersion)

FBartos replied:
yes, you are right. I thought that we had updated that but it must've gotten lost somewhere. Thanks for pointing it out.

FBartos replied:
thanks for noticing the errors in the help files, I updated them with one of the commits

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -15,408 +15,584 @@
+          return(e)
+      )
+    } else if (type == "GLMM") {
+      # needs to be avaluated in the global environment
+      glmm_family <<- options$family
+      glmm_link   <<- options$link
```
there are some weird problems with scoping and passing some input to afex which then passes it forward to glmer and etc. If you just run the code in a normal R-session, it works well with normal assignment, however, JASP returns an error and says that the variable doesn't exist. I encountered this problem in some other places and this was the only thing that managed to make it work.

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -15,408 +15,584 @@
+      
+      # I wish there was a better way to do this
+      if (options$family == "binomial_agg") {
+        glmm_weight <<- dataset[, .v(options$dependentVariableAggregation)]
```
same problem as described above, all `<<-` are used in cases with some scoping issues

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -15,408 +15,584 @@
+      if (any(attr(model, "class") %in% c("std::runtime_error", "C++Error", "error"))) {
+        if (model$message == "(maxstephalfit) PIRLS step-halvings failed to reduce deviance in pwrssUpdate") {
+          ANOVAsummary$setError(
+            "The optimizer failed to find solution. Probabably due to quasi-separation in the data. Try removing some of the predictiors."
```
oh OK, thanks. didn't know that. I will change it everywhere apropriatelly

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -15,408 +15,584 @@
-      effect_name <- "Intercept"
-    }else{
-      effect_name <- paste(.unv(unlist(strsplit(rownames(model$anova_table)[i], ":"))),collapse = ":")
+    ANOVAsummary$addColumnInfo(name = "effect",
+                               title = "Effect",
+                               type = "string")
```
that's what the ctrl+shift+a code "beautifier" did...., I don't like the visuals of the code at all now

vandenman replied:
I don't really like what Rstudio does... You could try `formatR` with `tidy_eval(file)` but somehow there's always something I don't like about it. Nevertheless, I can live with this regardless of how odd it looks.

FBartos replied:
the tide_eval doesn't work for me, it drops arguments from some functions. So, we are stuck with JASP-compliant ugly code

vankesteren commented:
Yeah, if that is by today I have time for it! Otherwise it's next week for me.

Also, I just remembered that maybe @JorisGoosen has something to say about the fact that this PR reverts the `description.qml` to a `description.json`?

FBartos commented:
I won't be able to finish it today. so a next week probably.

vandenman commented:
I'm happy to take another look at it, both this or next week!

JorisGoosen commented:
Yeah the module should have a qml not a json, sorry about that!

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -15,408 +15,584 @@
+      if (x$randomSlopes) {
+        return(NA)
+      } else{
+        return(gsub(" ", "", .v(unlist(x$value)), fixed = TRUE))
```
it's actually a remnant, the qml component was adding an empty space in to the variables if there was an interaction, some later version apparently fixed it - so it's not needed anymore. I remove the gsub now.

FBartos commented:
Hi @vandenman & @vankesteren ,

I think that I addressed all of the issues in the previous commits. I wasn't able to reproduce the crash, but I did encounter a similar one which should be fixed by now (https://github.com/jasp-stats/jasp-desktop/pull/4116#event-3412079012).  Could you please take another look at the module?

Thanks

vandenman commented on R/MixedModelsCommon.R:
```diff
@@ -15,408 +15,651 @@
+    family_text <- substr(family_text, 1, nchar(family_text) - 1)
+    if (options$family %in% c("Gamma", "inverse.gaussian", "neg_binomial_2", "poisson")) {
+      if (any(dataset[, .v(options$dependentVariable)] < 0))
+        stop(gettextf("%s requres that the dependent variable is > 0.",family_text))
+    } else if (options$family %in% c("binomial")) {
+      if (any(!dataset[, .v(options$dependentVariable)] %in% c(0, 1)))
+        stop(gettextf("%s requres that the outcomes are only 0 and 1.",family_text))
+    } else if (options$family %in% c("betar")) {
+      if (any(dataset[, .v(options$dependentVariable)] < 0))
+        stop(gettextf("%s requres that the outcomes is higher than 0 and lower than 1.",family_text))
```
`JASP:::.quitAnalysis` rather than `stop`.

FBartos replied:
thanks, didn't know about this function before

FBartos commented:
Thanks Don for finding the errors.

> I can enter non-scale variables (e.g., contBinom) but unlike linear regression, the analysis throws an error. For the sake of consistency, either prohibit non-scale variables or "try" to work for non-scale data as well by converting it to numeric.

I don't think that the conversion is a good idea, so I disabled the option for LMM/BLMM. However, it's not possible to do so for GLMM/BGLMM, because the `allowedColumns` is a static element and Bruno said that he couldn't change it in time for the release.

> the analysis doesn't like odd variables very much

I addressed this by adding more variable checks.

> the column Variance in the table obtained by checking Variance/correlation estimates looks off

yes, you are correct. This is very embarrassing. It has been consistently wrong thorough the analysis.

> Error handling for the link function could be improved

I added additional checks for the families and converted them into the `JASP:::.quitAnalysis`
